docs/design/phases.md

```diff
@@ -0,0 +1,354 @@
+     - `--organization=""`
+   - Output:
+     - The kubeconfig is `stdout`
+ - `kubeadm phase kubeconfig token`
+   - Inputs:
+     - `--client-name=`
+     - `--api-servers=`
+     - `--token`
+     - `--cert-dir=/etc/kubernetes/pki`
+   - Output:
+     - The kubeconfig is `stdout`
+ - `kubeadm phase controlplane etcd`
+   - Inputs:
+     - `--image=gcr.io/google_containers/etcd-${GOARCH}:3.0.14-kubeadm`
+     - `--data-dir=/var/lib/etcd`
+   - Output:
+     - `etcd.yaml` in the manifests dir (`/etc/kubernetes/manifests`)
+ - `kubeadm phase controlplane apiserver`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-apiserver-${ARCH}:${STABLE_VERSION}`
+     - `--service-cidr=10.96.0.1/12`
+     - `--cert-dir=/etc/kubernetes/pki`
+     - `--port=6443`
+     - `--address=0.0.0.0`
+     - `--etcd-servers=http://localhost:2379`
+     - `--authorization-mode=RBAC`
+     - `--extra-args`
+     - `--extra-volumes`
+   - Output:
+     - `kube-apiserver.yaml` in the manifests dir (`/etc/kubernetes/manifests`)
+ - `kubeadm phase controlplane controller-manager`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-controller-manager-${ARCH}:${STABLE_VERSION}`
+     - `--kubeconfig=/etc/kubernetes/controller-manager.conf`
+     - `--pod-network-cidr`
+     - `--cert-dir=/etc/kubernetes/pki`
+     - `--auto-approve-csrs=true`
+     - `--extra-args`
+     - `--extra-volumes`
+   - Output:
+     - `kube-controller-manager.yaml` in the manifests dir (`/etc/kubernetes/manifests`)
+ - `kubeadm phase controlplane scheduler`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-scheduler-${ARCH}:${STABLE_VERSION}`
+     - `--kubeconfig=/etc/kubernetes/scheduler.conf`
+     - `--extra-args`
+   - Output:
+     - `kube-scheduler.yaml` in the manifests dir (`/etc/kubernetes/manifests`)
+ - `kubeadm phase addons cluster-info`
+   - Inputs:
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+   - Outputs:
+     - Creates the `cluster-info` ConfigMap in the `kube-public` namespace
+     - Exposes it publicly with RBAC bindings
+- `kubeadm phase addons self-hosted`
+   - Inputs:
+     - Assumes the control plane yaml manifests are in `/etc/kubernetes/manifests`
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+     - `--cert-dir=/etc/kubernetes/pki`
+   - Outputs:
+     - Creates a selfhosted control plane from static pod manifests
+- `kubeadm phase addons kube-proxy`
+   - Inputs:
+     - `--image=gcr.io/google_containers/kube-proxy-${ARCH}:${STABLE_VERSION}`
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+     - `--pod-network-cidr=`
+   - Outputs:
+     - Applies the built-in `kube-proxy` manifest to the API Server
+- `kubeadm phase addons kube-dns`
+   - Inputs:
+     - `--admin-kubeconfig=/etc/kubernetes/admin.conf`
+     - `--dns-domain=cluster.local`
+     - `--image-repository=gcr.io/google_containers`
+   - Outputs:
+     - Applies the built-in `kube-dns` manifest to the API Server
+
```

Comment:
I suggest to provide a more consistent approach for outputs across the entire `kubeadm phase` UX, harmonising current approaches:
- `kubeadm phase certs selfsign` use `--cert-dir=`
- `kubeadm phase kubeconfig clientcert` writes to `stdout`
- `kubeadm phase controlplane *` writes to `/etc/kubernetes/manifests/*`
- `kubeadm phase addons cluster-info` apply directly to the apiserver
- eg

A first draft for possible alternative solution is:
- by default all `kubeadm phase` commands should provide the same experience of `kubeadm init`:
  - `kubeadm phase certs` and `kubeadm phase token` writes to `--cert-dir`, which defaults to `/etc/kubernetes/pki`
  - `kubeadm phase kubeconfig clientcert` writes to `/etc/kubernetes/; to be defined a rule for deriving filename from `--client-name=`(this is easier to implement if we will specialize kubeconfig phase)
  - `kubeadm phase controlplane *` writes to `/etc/kubernetes/manifests/*.yaml`
  - `kubeadm phase addons *` apply directly to the apiserver
- if invoked with a `--stdout`flag, all `kubeadm phase` commands should write to stdout; the output for the most of the commands is a yaml string or a sequence of kubectl commands; to be defined a format for `kubeadm phase certs`